### DISCUSSION TITLE
Raspberry with AlsaEQ and Softvol

### DISCUSSION THREAD

**Discussion by Dangoo** [mikebrady/shairport-sync#1293] [Show and tell]
Hey everyone,

first of all I'd like to give a huge shout out to @mikebrady and the other authors of this library and everyone who helped to make AP2 happen!

## Rationale

I was waiting for this to happen for quite some time to get started with a personal project to built Raspberry Pi based multiroom speakers that match my budget as well as my taste.

So I built the first prototype using these components:
- Raspberry Pi Zero W
- Adafruit I2s Bonnet
- Visaton FR 10 HM 4 Ω broad band speaker

To fine-tune the audio to optimize for the physical context of the speaker I searched for a simple EQ solution and found there's already the native Alsa EQ which was just what I needed for a beginning.
As I have good knowledge and experience in scripting languages but neither in development for real time audio nor in acoustic engineering, most of my findings and workarounds are based on reading long forum posts and try and error. So if you have any feedback, pls don't hesitate to let me know 😀 

**TL;DR**: Most of the issues could be solved by changing shairport-syncs user to `pi` 🤔 

So to get started you have to install plugin equal  - *dough*
There was a simple guide for HiFiBerry:
https://www.hifiberry.com/docs/software/guide-adding-equalization-using-alsaeq/

```bash
sudo apt-get install -y libasound2-plugin-equal
```

Fine, then some config:
```properties
# Passing default output to equalizer plugin
pcm.!default {
  type plug
  slave.pcm plugequal;
}
ctl.!default {
  type hw card 0
}
ctl.equal {
  type equal;
}

# Add equalizer plugin and pass output to hardware
pcm.plugequal {
  type equal;
  slave.pcm "plughw:0,0";
}
pcm.equal {
  type plug;
  slave.pcm plugequal;
}
```

Rebooting the Pi sadly let to no audio at all when playing from shairport-sync. On another topic here in the issues I found the [advice that shairport-sync is executed by its own user](https://github.com/mikebrady/shairport-sync/issues/1046), while `alsamixer -D equal` runs on default user. So adding this to the config finally made it work:
```properties
# /etc/systemd/system/shairport-sync.service
[Service]
User=pi
Group=pi
```

In order to prevent the Adafruit I2s Bonnet to produce a loud "popping" noise each time playback started a similar approach was needed. And once this was figured out, adding `softvol` which is required by the Adafruit Bonnet wasn't that tricky at all anymore.

So I collected my findings in a template to have it at hand for the following generation of speakers:
https://github.com/Dangoo/raspberry-audio-config

## The Question

Which now brings me to the question whether I got smth terribly wrong or is there maybe a better way to properly use Alsa EQ with shairport-sync than playing with users?

----

And to whom it may concern here's the working prototype with 3D printed housing - design is inspired by famous [Braun SK 1](https://www.radiomuseum.org/r/braun_kleinsuper_sk1.html?language_id=2):

![speaker front side](https://user-images.githubusercontent.com/1252204/134032015-215e7b88-cd68-4ad2-878e-fad33f7c3976.jpg)

![speaker backside and interior](https://user-images.githubusercontent.com/1252204/134032054-372512e5-83f5-4136-88d5-62a1119a9bd4.jpg)

**Comment by mikebrady** [comment]
That looks great — congratulations!

Your approach is fine, but there is an alternative, which is to use the built in equalizer. You can find a guide in the wiki. The potential advantages are that the amount of processing and transcoding should be less and the output should be fully synchronized. But TBH, you may not hear much of a difference.

**Comment by noelhibbard** [comment]
Thanks for shairing!

I love that enclosure. Did you print it yourself or can you buy one somewhere?
